allennlp/interpret/attack/input_reduction.py

```diff
@@ -0,0 +1,123 @@
+from typing import Dict, List, Set
+import numpy as np
+import torch
+from allennlp.interpret.attack import Attacker
+from allennlp.common.util import JsonDict, sanitize 
+from allennlp.data import Instance
+from allennlp.predictors.predictor import Predictor
+from collections import defaultdict
+
+@Attacker.register('input-reduction')
+class InputReduction(Attacker):
+    """
+    Runs the input reduction method from https://arxiv.org/abs/1804.07781.
+    """
+    def __init__(self, predictor):
+        super().__init__(predictor)
+
+    def attack_from_json(self, inputs:JsonDict, target_field: str, gradient_index:str,ignore_tokens:List[str] = ["@@NULL@@"]):       
```

Comment:
This one also should be broken up as much as possible.  It's possible that in doing this you'll find some logic that can be moved to the base `Attacker` class, though I wouldn't try too hard if there's nothing obvious.